src/contract/link.rs

```diff
@@ -0,0 +1,470 @@
+    }
+
+    /// Adds a library to deploy.
+    pub fn deploy_library<L>(self) -> Self
+    where
+        L: DeployableLibrary,
+        I: DependsOn<L>,
+    {
+        self.deploy_library_bytecode(L::name(), L::bytecode().clone())
+    }
+
+    /// Adds a library to deploy by name and bytecode.
+    pub fn deploy_library_bytecode<S>(self, name: S, bytecode: Bytecode) -> Self
+    where
+        S: AsRef<str>,
+    {
+        self.add_library(name.as_ref(), Library::Pending(bytecode))
+    }
+
+    /// Add a library to the current dependency graph.
+    ///
+    /// Note that this method always succeeds, this is because we do not link
+    /// incrementally, but rather when `build` is called so that linking errors
+    /// only need to be handled in one place.
+    fn add_library(mut self, name: &str, library: Library) -> Self {
+        let name = name.to_owned();
+        self.libraries.push((name, library));
+        self
+    }
+
+    /// Links the libraries and binaries together and returns a verified
+    /// deployment.
+    ///
+    /// This method will return an error if it finds unresolved or unused
+    /// libraries during the linking process.
+    pub fn link(mut self) -> Result<Deployment, LinkerError> {
+        let mut included_libraries = HashSet::new();
+        let mut pending_libraries = HashMap::new();
+        for (name, library) in self.libraries {
+            if !included_libraries.insert(name.clone()) {
+                return Err(LinkerError::DuplicateDependency(name));
+            }
+
+            match library {
+                Library::Resolved(address) => self.contract_bytecode.link(&name, address)?,
+                Library::Pending(bytecode) => {
+                    pending_libraries.insert(name, bytecode);
+                }
+            }
+        }
+
+        let mut libraries_to_deploy = Vec::new();
+        for library in self.contract_bytecode.undefined_libraries() {
+            if let Some((name, bytecode)) = pending_libraries.remove_entry(library) {
+                let bytes = match bytecode.to_bytes().ok() {
+                    Some(bytes) => bytes,
+                    None => return Err(LinkerError::NestedDependencies(name)),
+                };
+                libraries_to_deploy.push((name, bytes));
+            } else {
+                return Err(LinkerError::MissingDependency(library.to_owned()));
+            }
+        }
+
+        // NOTE: At this point, the contract bytecode should be completely
+        //   linkable, as we linked all the library instance addresses and
+        //   verfied that the remaining dependencies are to be deployed. The
+        //   libraries remaning in `pending_libraries` are extra uneeded
+        //   dependencies since we already removed the libraries that were
+        //   required for linking from the hash map. Report an error with the
+        //   first unused dependency.
+        if let Some(unused_dependency) = pending_libraries.keys().next() {
```

Comment:
I'm wondering if returning an error here is the best behavior. Imagine you have a contract that has unused dependencies for whatever reason. You wouldn't be able to use ethcontracts with it. I'm not sure if this could happen in practice or is really an invalid state, but maybe we should just ignore those libraries?

Reply:
The unused dependency here is a dependency that was specified to the linker but not needed by the contract (as in it has no place holder).

Note that the generated code should only ever permit libraries that are needed by the contract bytecode (by the presence of the placeholders in the bytecode) so this shouldn't be a problem for the generated code. 

To be honest, I'm not sure I understand the scenario you are mentioning.

Reply:
So you are saying this input is controlled by the user of ethcontract? It's not taken e.g. from the artefact json that they might be getting from some external smart contract they might have no control over?